### PR TITLE
feat(log-injection): example vulnerable code for CSRF injection

### PR DESCRIPTION
**Pull request by lirantal**
Adding an example of a Log Injection that can result in log forging/tampering if user input is logged unsanitized.

Vulnerable code also has comments about how to fix the problem.

@ckarande what do you think?

### CODE REVIEW

**Comment by lirantal** [comment]
pushed tutorial description now too

**Comment by binarymist** [comment]
Haven't tested this, but love the concept. Definatly something devs need to be thinking about. I'm aware that devs like an out of the box solution to these things, and don't like to think to much about security (currently), I see you mention context in several places... Possibly some more comments making it even more explicit that it really depends on execution context as to how the sanitisation should be performed. One method does not suite all... sort of thing. Devs really need to understand where the untrusted data will flow and sanitise according to execution contexts.

Also worth thinking about constraining untrusted data to domain specific types (validation, filtering) long before it gets to any sanitisation (Just thinking out loud).

**Comment by lirantal** [comment]
Thanks for the feedback. I agree on the context and it's generally already covered in output handling  but I also wanted to make it clear so there are several code examples on how to encode it to different execution contexts.

File: app/routes/session.js [hunk 1]
```diff
@@ -61,6 +61,20 @@ function SessionHandler(db) {
             var invalidPasswordErrorMessage = "Invalid password";
             if (err) {
                 if (err.noSuchUser) {
+                    console.log('Error: attempt to login with invalid user: ', userName);
+
+                    // Fix for A1 - 3 Log Injection - encode/sanitize input for CSRF Injection
```
Comment (ckarande) [inline app/routes/session.js]: Would this be  a CSRF or CRLF Injection?

File: app/routes/session.js [hunk 2]
```diff
@@ -61,6 +61,20 @@ function SessionHandler(db) {
...
+                    // that could result in log forging:
+                    // - Step 1: Require a module that supports encoding
+                    // var ESAPI = require('node-esapi');
+                    // - Step 2: Encode the user input that will be logged in the correct context
+                    // following are a few examples:
+                    // console.log('Error: attempt to login with invalid user: %s', ESAPI.encoder().encodeForHTML(userName));
+                    // console.log('Error: attempt to login with invalid user: %s', ESAPI.encoder().encodeForJavaScript(userName));
+                    // console.log('Error: attempt to login with invalid user: %s', ESAPI.encoder().encodeForURL(userName));
+                    // or if you know that this is a CSRF vulnerability you can target this specifically as follows:
```
Comment (ckarande) [inline app/routes/session.js]: Similar to above comment, do you mean CRLF instead of CSRF here?

File: app/routes/session.js (hunk 1 above)
Comment (lirantal) [reply app/routes/session.js]: You're totally right on the CSRF thing - I meant to write [CRLF](https://www.owasp.org/index.php/CRLF_Injection) but my brain autocompleted the wrong thing :-)

File: app/routes/session.js (hunk 2 above)
Comment (lirantal) [reply app/routes/session.js]: Yep, as above, meant CRLF.

**Comment by lirantal** [comment]
In the comments we can keep CRLF because that's the actual specific vulnerability we demo in the app, and the tutorial has a "Log Injection" title to explain the matter in a more generic way.

Fixed and ready from my side.

**Pull request merged**